GOVERNANCE.md

```diff
@@ -0,0 +1,150 @@
+
+* **Team members**: Any members of the private [prometheus-team][team] Google group.
+
+* **Maintainers**: Team members who are leads on an individual project ([MAINTAINERS.md][maintainers.md]).
+
+* **Projects**: Any repository in the Prometheus [GitHub organization][gh].
+
+* **The Prometheus Project**: The sum of all activities performed under this governance, concerning one or more Repositories or the community.
+
+## Values
+
+The Prometheus developers and community are expected to follow the values defined in the [CNCF charter][charter], including the [CNCF Code of Conduct][coc]. Furthermore, the Prometheus community strives for kindness, giving feedback effectively and building a welcoming environment. The Prometheus developer community strives to operate by rough consensus for the majority of decisions and only resort to conflict resolution through a motion if consensus cannot be reached.
+
+## Projects
+
+Each project must have a MAINTAINERS.md file with at least one maintainer. Where a project has a release process, access and documentation should be such that more than one person can perform a release. Releases should be announced on the prometheus-users mailing list. Any new projects should be first proposed on the [developers mailing list][devs] following the voting procedures listed below. When a project is no longer relevant, it should be moved to the prometheus-junkyard Github organization.
+
+## Decision making
+
+### Team members 
+
+Team member status may be given to those who have made ongoing contributions for at least 3 months to the Prometheus Project. This is usually in the form of code improvements and/or notable work on documentation, but organizing events or user support could also be taken into account. 
+
+New members may be proposed by any existing member by email to [prometheus-team][team] and voted on by [lazy consensus](#consensus). If no consensus is reached, the new member is not accepted.
+
+Team members are added to the [GitHub organization][gh] and have commit rights to all projects. They should however respect the maintainers of each project.
+
+Team members may retire at any time by emailing [the team][team].
+
+Team members can be removed by [supermajority vote](#supermajority-vote) on [the team mailing list][team]. For this vote, the member in question is not eligible to vote and does not count towards the quorum.
+
+Upon death of a member, their team membership ends automatically.
+
+Unless voted upon otherwise, membership changes take effect immediately.
+
+A public list of team members is kept on the [community page][community] or linked from it.
+
+### Maintainers
+
+Maintainers are team members who lead one or more project(s) and serve as a point of conflict resolution amongst the contributors to this project.
```

Comment:
We already concluded that team members don't necessarily need to be maintainers. We have only one case of the inverse, but I guess it's in general not a problem to be a maintainer without being a team member.

Perhaps we should clarify that both cases are possible? We should also mention that maintainers that are not team members still have commit rights to all projects (don't they?).

Reply:
> We have only one case of the inverse, but I guess it's in general not a problem to be a maintainer without being a team member.

It's a bit messy, I'd like if maintainers were a subset of team.

> We should also mention that maintainers that are not team members still have commit rights to all projects (don't they?).

They don't currently, only -team.

Reply:
> It's a bit messy, I'd like if maintainers were a subset of team.

That was also my understanding. That implies adding the one person who is only a maintainer to the team.

Reply:
For reference: https://github.com/prometheus/prometheus_api_client_ruby/blob/master/MAINTAINERS.md

However, I remember a branch of the discussion where some people said they would not see Yaacov's involvement as sufficient to enter -team.

Reply:
If anyone is still of that opinion, please speak up. My understanding was that the consensus evolved away from that to eliminate a lot of edge cases.

Reply:
I agree that "all maintainers are also team members" simplifies things. But perhaps the case where somebody marginally contributes to the project, needs push rights for that, but doesn't even want the responsibility of voting is a real thing.

Reply:
I don't think we got much discussion on that particular point. Maybe something to discuss on the list?

Reply:
> somebody marginally contributes to the project, needs push rights for that

that's not the same as maintainership to me. We don't have an explicit clause, but if we want that we can either make it so as an "other matter". Or handle it in this doc.

Reply:
> somebody marginally contributes to the project, needs push rights for that

Yeah, sorry. I meant (and should have written) "somebody _maintains_ a marginally important repo in the Prometheus org and needs push rights for that but..."

Reply:
And no, I don't think we need a role for somebody that is neither a maintainer nor a team member but needs push rights. I think it is fine that you need to be either a maintainer or a team member to gain push rights.

Reply:
Sounds like we're in agreement on that point. How about all maintainers needing to be in -team?

Reply:
I thought that was implicitly part of the vote on prometheus-team about whether the voting base is prometheus-team or maintainers, and one of the worries would be that if it's maintainers, we'll have to put in place more stringent procedures for adding new maintainers.

Thus I think I'm against automatically adding every maintainer to prometheus-team.

Reply:
(to clarify, yes, I know that was not *all* that vote was about - it was also about non-maintainer team members who have a big enough impact that they should be able to vote)

Reply:
I'd see it as the other way around, to be a maintainer you need to be on prometheus-team first.

Reply:
So yes, in case I didn't express myself clearly, I do believe it is useful to allow maintainers that are not in prometheus-team (and vice versa). I think there is very little that we need to tweak in the current governance doc to accommodate the case (essentially just that maintainers have GH push permissions to (perhaps not ownership?) and to clarify that membership in either role is independent of the other.

Reply:
I'm kinda against that as it makes our permissions model more complicated, and it seems odd if we're having a discussion on -team around e.g. 3.0 timeline that not all maintainers would be privy to that.

Reply:
I would also prefer to keep them the same, for the sake of communication, and for the sake of saving on governance clauses … but if the majority thinks otherwise, what would you want the prerequisites to be? (explicit consensus, simple vs. supermajority?)

Reply:
I agree it makes the permissions model more complicated. My point is that it is a necessary complication.

About your other point: First, I think that a discussion around the 3.0 timeline should happen or prometheus-developers@. Second, I think a maintainer that is not on prometheus-team (the only practical example right now is Yaacov, maintaining https://github.com/prometheus/prometheus_api_client_ruby) is probably not so much involved in core decisions to be interested in things like the 3.0 timeline. That's kind of the point of that "maintainer only" role.

Reply:
@matthiasr I assume with "the same" you mean maintainers should be a *subset* of prometheus-team, right? (my opinion stands...)

Reply:
Yes, that's what I meant.

Reply:
If someone is maintaining something but isolated from prometheus-team, I'd start wondering why what they're maintaining is inside the Prometheus org as against the 200+ integrations that are outside the org.

Reply:
We have a concrete example: https://github.com/prometheus/prometheus_api_client_ruby
Should we…
* …move the repo to a place outside of the Prometheus org?
* …make Yaakov a prometheus-team member?
* …find a different maintainer?

Reply:
I did some digging on this one concrete case of the Ruby API client lib and how it came about. Yaacov from RedHat built this and was interested in putting it into the Prometheus org. An email from @grobie back then to Yacoov read:

> I created a poll for the Prometheus maintainers team to see if anyone has objections to merge it into the Prometheus main org. I expect a result this weekend. In case there are no major objections, I'll get back to you about the exact steps to transfer ownership.
> 
> Just to double check, you/RedHat will continue to be a maintainer of that API client and help out with user requests and features, correct?"

You (@brian-brazil) ended up saying (in different thread):

> The library in general seems new, unused and a bit low level so will require development. As long as you're happy to take on the maintenance burden, I see no problems.

Thus it ended up in our org. But: I have nothing to say for or against Yaacov, I just think someone should be more well-known to and involved with the project and community before becoming a voting full team member. Otherwise we also risk filling up the voting pool with people who don't follow along with discussions/voting and thus make reaching quorums harder.

What should we do in this concrete case? Add him to team anyway? Remove him as a maintainer? Or have maintainers / prometheus-team be separate (but highly correlated) lists? I'm for the latter...

Reply:
> Otherwise we also risk filling up the voting pool with people who don't follow along with discussions/voting and thus make reaching quorums harder.

I have the same concern with filling up the org with people not on -team. That could set us down a path where a large swathe of the integrations ended up in the org, which is not something -team have bandwidth to manage. Merely writing an integration should not be enough to get code in the org nor to become a member -team.

My preferences are in order to a) list grobie as maintainer b) remove this repo from the org (if noone in -team is willing to maintain it...) c) add him to -team.

Reply:
Another perspective on the same thing: Whenever we discussed onboarding more contributors, a common concern was that we have plenty of contributions but we need more people that will take some responsibility. The role of a maintainer of a smaller repo or a part of a larger repo is a nice graduation path to become a full prometheus-team member. By requiring prometheus-team membership for each maintainer (or even for everybody with push permissions), we put the bar very high (in particular if we require explicit consensus). Making somebody a maintainer without prometheus-team membership can be pretty lightweight, though. Since it is easy to revert (including any vandalism somebody might have done using their push permissions), we can rely on lazy consensus here.

Reply:
That's a really good point. I think we can still be conscious about adding new projects, this is already handled in this doc. I'll make another pass on this doc and can amend the maintainership section to account for this.

Should someone who maintains a repo but is not on the team get push access everywhere or not?

Reply:
Another practical example: https://github.com/prometheus/client_golang/tree/master/api . This is effectively abandonware. I'm formally the maintainer as I am the maintainer of the repo, but this directory ended up there before we formalized that, and I always insisted that I'm currently not in a position to maintain this part of the repo. Nobody already in prometheus-team stepped in. But we get the occasional contribution for it, which is then quite awkward because somebody has to review it and make the call. With @brian-brazil 's preferences above, we needed to remove the API from the Prometheus GH org, as we have already concluded that nobody in -team is willing to maintain it.

The trajectory I envision is more the following: It would be very natural to ask one of those regular contributors to become the maintainer, e.g. @andrestc, but it would be a bit much to make him a fully voting member with the limited interactions we had with him so far. If the API part of the repo remains in a state of rare changes, it's fine to keep it that way. We have gained one semi-active maintainer. Win! If André happens to get more involved, it will be very natural again to graduate to full -team membership. We have gained another full team member. Win again!

With the "big bang" requirement, it will be much more unlikely to ever reach that state.

Reply:
> Should someone who maintains a repo but is not on the team get push access everywhere or not?

Not being a true expert in GH permissions, but I'd tend to keep it as simple as possible and grant push access everywhere, but not ownership of the org (which would be something we only grant to -team members).

Reply:
> I have the same concern with filling up the org with people not on -team.

I think we should state somewhere that being a maintainer without being a -team member is not meant to be common. It is either for special cases or for the suggested graduation path, in which it is a transition rather than a desired end state.

Reply:
The way I'm imagining things is that an existing -team member would mentor a potential new maintainer/member, and be the one to merge PRs until they became a member of -team and could become maintainer. This is what we already roughly do elsewhere where there's a regular and upcoming contributor to code that already has a clear maintainer.

Reply:
Yes, I get that. But that's also what I mean: It's a bumpier entry path. It requires a lot more explicit commitment. If it works out, great. But in practice, I see all of us just being busy enough to not commit to mentorship for a possible future maintainer of a perhaps not that important library like the Ruby or Go API client.

Reply:
If we have a maintainer who can't actually merge anything, what exactly is their role? In this scenario, the real maintainer is the one "mentoring", and they are in no different a position than any other commenter/contributor.

I feel strongly that if we entrust someone with the maintenance of something, they should have the means to do that.

All the essentially-unmaintained code you mentioned has been contributed but the people who have an interest in maintaining it cannot do so.

Reply:
> If we have a maintainer who can't actually merge anything, what exactly is their role? In this scenario, the real maintainer is the one "mentoring", and they are in no different a position than any other commenter/contributor.

I think we're saying the same thing, from different perspectives. Such a person is not a maintainer, but they can be a mentee.

> All the essentially-unmaintained code you mentioned has been contributed but the people who have an interest in maintaining it cannot do so.

I don't think it's getting to the stage of cannot, as there is no maintenance being attempted. No PRs, no code reviews, or comments on issues. If there is maintenance happening on an ongoing basis, getting into -team shouldn't be difficult from there.

Reply:
Meta-level remark: we don't have to get all aspects of the governance perfect in the first PR + vote. We can still discuss and vote on amendments around some of these finer points after we get the 1.0 accepted. But then I'd recommend keeping it conservative initially (e.g. not adding every maintainer into prometheus-team) and put to later discussion + vote whether we want to open things up more.

Reply:
> What if it doesn't work out? We've seen code added abandoned a fair bit, SD providing several examples of that. I think it'd be better to not give out permissions too early than to have to revoke them later on.

The alternative to a bar too high is not necessarily a bar too low, it _can_ be a bar at just the right level. My point is that getting into -team is a bar too high. Your point is that it is just the right level. Not sure if we can it dissect anymore at this point.

Reply:
Overall, I think this particular aspect is more a detail than a huge deal. How about we take an informal vote here to see what the current mood is instead of discussing it in even more detail? If, after all those explanations above, there is a clear majority either way, we can go with it for the initial version of the governance.

Thus: Please :+1: this comment if you are _for_ the (rare) possibility of maintainers with push rights not being part of prometheus-team. Please :-1: this comment if you think maintainers must always be part of prometheus-team.